work in progress

jwong-nd opened the pull request:
Initial ideas on curriculum

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+        TaskId = int
+        # ID tasks on Task.name, a required field.
+        # Tasks are allowed to change parameters.
+        self.tasks: dict[str, TaskId] = {}
```
A curriculum is not a collection of tasks, it's a collection of stages. In theory you could have the same task twice. Mostly just a nomenclature clarification.

alexpiet replied:
Where did we land on whether we need a schema for a stage?

bruno-f-cruz replied:
I think I am almost convinced depending on how we deal with the modifiable attritube tag and assuming we followed the proposal that a stage is a stage of a task iif you can validate the stage instance against the task definition.

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+        TaskId = int
+        # ID tasks on Task.name, a required field.
+        # Tasks are allowed to change parameters.
+        self.tasks: dict[str, TaskId] = {}
```
Why pass taskid instead of the Task object itself?

bruno-f-cruz replied:
The two collections should not have setters. The collections should not be modifiable using the add/pop methods otherwise it will be too easy to break the graph.

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+        # The collection is ordered by order of Curriculum API calls.
+        self.graph: dict[TaskId, dict[TaskId, Evaluation]] = defaultdict(dict)
+
+    def add_node(self, task: Task):
```
Use the domain language. Eg add_stage seems more adequate.

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+        self.tasks: dict[str, TaskId] = {}
+        # Each node references a collection of (edge, destinatation node) associations.
+        # The collection is ordered by order of Curriculum API calls.
+        self.graph: dict[TaskId, dict[TaskId, Evaluation]] = defaultdict(dict)
```
What is the advantage of having the nested dictionary? Shouldn't the value just be an array of Tuple(destination_stage, eval)? 
I can see a world where the keys will not be unique. For instance the same stage with multiple rules. Also, the order the this list matters so I think this is better suited to be an Array?

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+    """
+    pass
+
+class Metric:
```
I think this class should not wrap the database. This should only spec what the metrics oughta look like. Remember I may want to use Metrics without using the database.

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+    def add_node(self, task: Task):
+        self.tasks[task.name] = len(self.tasks)
+
+    def add_connection(self, start_task: str, next_task: str, eval: Evaluation):
```
add_transition

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+
+
+# TODO: De/serialization
+class Curriculum:
```
I think we will need pop/del methods too. For instance, if you pop a stage you want to make sure you tell the user that some connections will have to be removed too. Also if you add a transition one should check if the stages are available too. 

For the transitions we will also need methods to order the evaluation of the rules. Since the order matters

bruno-f-cruz replied:
We might also need a "finally" -like evaluation that runs if no transition evaluates to true. This clause is responsible for updating the current stage settings instead of transitioning the animal. Should probably be a dict[stage, eval] since there should only be one per stage.

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+        pass
+
+
+def combine_curriculums(curr_1: Curriculum, curr_2: Curriculum) -> Curriculum:
```
Good idea!

bruno-f-cruz commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,80 @@
+    pass
+
+
+class Principal:
```
Love the class name!

alexpiet commented on src/aind_behavior_curriculum/tmp.py:
```diff
@@ -0,0 +1,106 @@
+        pass
+
+
+def combine_curriculums(curr_1: Curriculum, curr_2: Curriculum) -> Curriculum:
```
Combining curriculums is a useful feature. I think we will need to provide more precise inputs to combine curriculums. I can imagine it working something like "Connect curriculum A, node/stage X with curriculum B at node/stage Y"

bruno-f-cruz replied:
Not sure you need this as a function to combine the two curricula. You can always merge 2 curricula even if there are no transitions between them. At least they look fully composable to me. Sure you will have to define rules if you want to have a final curriculum that makes sense, but from the computational point of view i am not sure why it would be a problem. The exception here is what to do if the two curricula have the same task/stage key in the dictionaries. I can see two things happening:
1. Throw if the union of the keys is not unique
2. Find equal keys, and take the union of the transition rules
This could even be an optional parameter in the concat function (e.g. force_outter_join = True)

jwong-nd commented:
Some updates to the proposed curriculum/abstract-Trainer based on recent discussion. 
This is the simplest 'policy' implementation I can think of that does not introduce a whole matrix of stages. 
Work in progress, serialization to be added.

jwong-nd commented:
This Update (commit a5bc1d2):

- Floating Nodes: 
Updates Stage and Curriculum API’s with ‘add_policy’/‘remove_policy’ and ‘add_stage’/‘remove_stage’ methods respectively to handle floating policies and floating stages. 
This is a more natural interface for adding/removing a single node over optional rule/end-node arguments in ‘add_transition/remove_transition’.
Remove_transition is augmented with the ability to remove start or end nodes along with Rule removal. 

- Improved Rule Initialization: 
Rules (Policy, StageTransition, Policy Transition) are defined by passing in a callable instead of overriding __call__. 
The signature of the input callable is checked on Policy/StageTransition/PolicyTransition initialization with Pydantic validation. 
The updated rule definition serializes properly as before. 

- Trainer Test Suite:
Trainer passes a minimal test suite of core functionality. This can be added upon.  

- Other minor changes:
    - Use of Pydantic model_copy and updated Rule interface in examples
    - Remove Metrics from Curriculum, which are Task-specific
    - Removed Inbound/Outbound typing, in which all policies/stages inside self.policies and self.stages lists are all simultaneously inbound and outbound. 
    - Removed ‘import aind_behvaior_curriculum as abc’ in favor of ‘from aind_behavior_curriculum import (…)’ to avoid overloading of ‘abc’ with python’s abstract class library.
